### PR TITLE
#177 Linq Queries

### PR DESCRIPTION
**Pull request by KorolS**

### CODE REVIEW

**Comment by KorolS** [comment]
#177 Linq

File: HW5/Svetlana Koroleva/Task/Task/LinqSamples.cs [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        [Title("Where - Task 6")]
+        [Description("This sample return return customers validated on postalcode, region, phone")]
+
+        public void Linq006()
+        {
+            Regex notnumeric = new Regex(@"(\D)");
+            var customers =
+                 from c in dataSource.Customers
+                 where
+                 notnumeric.IsMatch(c.PostalCode, 0)
```
Comment (vladimirchr) [inline HW5/Svetlana Koroleva/Task/Task/LinqSamples.cs]: We need to check `c.Postalcode` for `null`
`c.PostalCode == null? false : notnumeric.IsMatch(c.PostalCode)`
                 
null

File: HW5/Svetlana Koroleva/Task/Task/LinqSamples.cs [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+        public void Linq008()
+        {
+            var avgPrice = dataSource.Products.Average(p => p.UnitPrice);
+            Dictionary<string, decimal> ranges = new Dictionary<string, decimal>();
+            ranges.Add("cheap", Math.Round(avgPrice * (decimal)0.67));
+            ranges.Add("average", Math.Round(avgPrice * (decimal)1.33));
+            ranges.Add("expensive", Math.Round(dataSource.Products.Max(p => p.UnitPrice)));
+
+            var grouped = dataSource.Products.Select(p => new { p.ProductName, p.UnitPrice }).OrderBy(p => p.UnitPrice).GroupBy(x => ranges.FirstOrDefault(r => r.Value > x.UnitPrice)).Select(gProduct => new { gProduct.Key, gProduct  });
```
Comment (vladimirchr) [inline HW5/Svetlana Koroleva/Task/Task/LinqSamples.cs]: Very interesting application of dictionaries. In my opinion
`decimal cheapThreshold = Math.Round(...);`
`decimal expensiveThreshold = Math.Round(...);`
`...`
`.GroupBy(p => p.UnitPrice < cheapThreshold ? "Cheap" :`
`p.UnitPrice > expensiveThreshold ? "Expensive" : "Middle price")...`
will be more readable and resource efficient. But it's just my opinion ;)

File: HW5/Svetlana Koroleva/Task/Task/LinqSamples.cs [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+                                ord.City,
+                                Totals=ord.Orders.Sum(o=>o.Total),
+                                Amount=ord.Orders.Count()};
+
+            var grouped = from q in sumbycity
+                          group q by q.City into cities
+                          select new {
+                              City = cities.Key,
+                              Totals=cities.Sum(c=>c.Totals),
+                              Amount= cities.Sum(c => c.Amount),
```
Comment (vladimirchr) [inline HW5/Svetlana Koroleva/Task/Task/LinqSamples.cs]: I think we need to change 
`Totals = cities.Sum(c=>c.Totals),`
`Amount = cities.Sum(c => c.Amount)`,
to
`Totals = cities.Average(c=>c.Totals),`
`Amount = cities.Average(c => c.Amount)`

File: HW5/Svetlana Koroleva/Task/Task/LinqSamples.cs (hunk 2 above)
Comment (KorolS) [reply HW5/Svetlana Koroleva/Task/Task/LinqSamples.cs]: I find my decision more suitable to print group names, I just like to use dictionaries)

File: HW5/Svetlana Koroleva/Task/Task/LinqSamples.cs (hunk 3 above)
Comment (KorolS) [reply HW5/Svetlana Koroleva/Task/Task/LinqSamples.cs]: Yes, I'm not right here

File: HW5/Svetlana Koroleva/Task/Task/LINQtoSQLModule.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        [Description("This sample returns customers with a turnover more than given value")]
+
+        public void Linq001()
+        {
+            decimal value = 10000;
+
+            var customers =
+                 from c in db.Customers
+                 where (c.Orders.Sum(o => o.Order_Details.Sum(t => (t.Quantity * t.UnitPrice * (decimal)(1 - t.Discount)))) > value)
+                 select new { c.CompanyName, Turnover = c.Orders.Sum(o => o.Order_Details.Sum(od => (od.Quantity * od.UnitPrice * (decimal)(1 - od.Discount)))) };
```
Comment (vladimirchr) [inline HW5/Svetlana Koroleva/Task/Task/LINQtoSQLModule.cs]: We could make variable for turnover
`from c in db.Customers`
`let turnover = c.Orders.Sum(o => o.Order_Details.Sum(t => (...)))`
`where (turnover > value)`
`orderby turnover descending`
`select new { c.CompanyName, Turnover = turnover };`

Comment (KorolS) [reply HW5/Svetlana Koroleva/Task/Task/LINQtoSQLModule.cs]: It's great idea! I will keep it in mind!

File: HW5/Svetlana Koroleva/Task/Task/LINQtoSQLModule.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        [Description("This sample return return customers validated on postalcode, region, phone")]
+
+        public void Linq006()
+        {
+           
+            var customers =
+                 from c in db.Customers
+                 where 
+                 
+                 System.Data.Linq.SqlClient.SqlMethods.Like(c.PostalCode.Trim(), "%[^0-9]")
```
Comment (vladimirchr) [inline HW5/Svetlana Koroleva/Task/Task/LINQtoSQLModule.cs]: US post codes has only 5 digits so we can use this expression to get all non valid postcodes:
`!SqlMethods.Like(c.PostalCode.Trim(), "[0-9][0-9][0-9][0-9][0-9]")`

Comment (KorolS) [reply HW5/Svetlana Koroleva/Task/Task/LINQtoSQLModule.cs]: And what about other countries? UK for example?

Comment (vladimirchr) [reply HW5/Svetlana Koroleva/Task/Task/LINQtoSQLModule.cs]: UK have non numerical post code system) Here is the postal code for London: SW15 5PU

File: HW5/Svetlana Koroleva/Task/Task/LinqSamples.cs [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+        [Title("Where - Task 8")]
+        [Description("This sample returns products grouped by ranges")]
+
+        public void Linq008()
+        {
+            var avgPrice = dataSource.Products.Average(p => p.UnitPrice);
+            Dictionary<string, decimal> ranges = new Dictionary<string, decimal>();
+            ranges.Add("cheap", Math.Round(avgPrice * (decimal)0.67));
+            ranges.Add("average", Math.Round(avgPrice * (decimal)1.33));
+            ranges.Add("expensive", Math.Round(dataSource.Products.Max(p => p.UnitPrice)));
```
Comment (spore9) [inline HW5/Svetlana Koroleva/Task/Task/LinqSamples.cs]: I like your approach in this task)

**Pull request merged**